rpcs3/Emu/RSX/Overlays/overlay_osk.cpp

```diff
@@ -29,37 +32,111 @@ namespace rsx
 			fade_animation.active = true;
 		}
 
-		void osk_dialog::initialize_layout(const std::vector<grid_entry_ctor>& layout, const std::u32string& title, const std::u32string& initial_text)
+		void osk_dialog::add_panel(const osk_panel& panel)
+		{
+			// On PS3 apparently only 7 panels are added, the rest is ignored
+			if (m_panels.size() < 7)
+			{
+				// Don't add this panel if there already exists one with the same panel mode
+				for (const auto& existing : m_panels)
+				{
+					if (existing.osk_panel_mode == panel.osk_panel_mode)
+					{
+						return;
+					}
+				}
+				m_panels.push_back(panel);
+			}
+		}
+
+		void osk_dialog::step_panel(bool next_panel)
+		{
+			const size_t num_panels = m_panels.size();
+
+			if (num_panels > 0)
+			{
+				if (next_panel)
+				{
+					m_panel_index = (m_panel_index + 1) % num_panels;
+				}
+				else if (m_panel_index > 0)
+				{
+					m_panel_index = (m_panel_index - 1) % num_panels;
+				}
+				else
+				{
+					m_panel_index = num_panels - 1;
+				}
+			}
+
+			update_panel();
+		}
+
+		void osk_dialog::update_panel()
 		{
+			ASSERT(m_panel_index < m_panels.size());
+
+			const auto panel = m_panels[m_panel_index];
+
+			num_rows = panel.num_rows;
+			num_columns = panel.num_columns;
+			cell_size_x = panel.cell_size_x;
+			cell_size_y = panel.cell_size_y;
+
+			update_layout();
+
 			const u32 cell_count = num_rows * num_columns;
 
 			m_grid.resize(cell_count);
+			num_shift_layers_by_charset.clear();
+
+			const position2u grid_origin = { m_frame.x, m_frame.y + 30u + m_preview.h };
+
 			u32 index = 0;
 
-			for (const auto& props : layout)
+			for (const auto& props : panel.layout)
 			{
 				for (u32 c = 0; c < props.num_cell_hz; ++c)
 				{
 					const auto row = (index / num_columns);
 					const auto col = (index % num_columns);
 					verify(HERE), row < num_rows && col < num_columns;
 
-					auto &_cell = m_grid[index++];
-					_cell.pos = { col * cell_size_x, row * cell_size_y };
+					auto& _cell = m_grid[index++];
+					_cell.button_flag = props.type_flags;
+					_cell.pos = { grid_origin.x + col * cell_size_x, grid_origin.y + row * cell_size_y };
```

Comment:
Why is this manual translation now necessary? Is there a missing container?

Reply:
Nvm, I see I went the cheap route with this myself earlier. I'll probably redo the elements packing logic later with a proper vertical layout container.